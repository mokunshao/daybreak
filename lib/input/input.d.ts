import React from 'react';
import './input.scss';
interface Props extends React.InputHTMLAttributes<HTMLInputElement> {
    clearable?: boolean;
    onClear?: Function;
}
declare const Input: React.FunctionComponent<Props>;
export default Input;
export { Input };
//# sourceMappingURL=input.d.ts.map