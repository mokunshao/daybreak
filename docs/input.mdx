---
title: Input
---

import { useState } from 'react'
import { Playground, Props } from 'docz'
import Input from '../src/input/input.tsx'

# Input

## Basic usage

<Playground>
  {()=>{
    const [value, setValue] = useState('');
    return (
      <Input value={value} onChange={(e) => setValue(e.target.value)} />
    );
  }}
</Playground>

## Clearable

`clearable` prop and `onClear` prop should be used together.

<Playground>
  {()=>{
    const [value, setValue] = useState('');
    return (
      <Input
        clearable
        value={value}
        onChange={(e) => setValue(e.target.value)}
        onClear={()=>setValue('')}
      />
    );
  }}
</Playground>

## Input for password

<Playground>
  {()=>{
    const [value, setValue] = useState('');
    const [value2, setValue2] = useState('');
    return (
      <div>
        <div>Basic:</div>
        <Input type="password" value={value} onChange={(e) => setValue(e.target.value)} />
        <div>Clearable:</div>
        <Input type="password" clearable value={value2} onChange={(e) => setValue2(e.target.value)} onClear={() => setValue2('')} />
      </div>
    );
  }}
</Playground>

## Props

<Props of={Input} />
