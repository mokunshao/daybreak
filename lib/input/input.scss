@import '../common.scss';

.daybreak-input {
  font-size: 1em;
  outline: none;
  padding: 0.5em;
  box-sizing: border-box;
  border: 2px solid transparent;
  background-color: $grey2;
  transition: all 200ms;

  &-wrapper {
    display: inline-flex;
    position: relative;
  }

  &:focus {
    border: 2px solid $black;
    background-color: $grey;
  }

  &-icons {
    position: absolute;
    top: 50%;
    right: 0.5em;
    transform: translateY(-50%);
    display: inline-flex;
    align-items: center;
  }

  &-icon {
    display: inline-flex;

    &-clear {
      cursor: pointer;
      width: 1em;
      height: 1em;
    }

    &-password {
      cursor: pointer;
      width: 1em;
      height: 1em;
    }
  }

  &-icon + &-icon {
    margin-left: 0.5em;
  }
}
