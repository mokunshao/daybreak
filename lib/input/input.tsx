import React, { useRef, useEffect, useState } from 'react';
import { classes } from '../utils/classes';
import { joinedClass } from '../utils/joinedClass';
import './input.scss';
import Icon from '../icon/icon';

const baseClass = joinedClass('input');

export interface Props extends React.InputHTMLAttributes<HTMLInputElement>{
  clearable?: boolean;
  onClear?: Function;
}

const Input: React.FunctionComponent<Props> = (props) => {
  const inputElement = useRef<HTMLInputElement>(null);
  const {
    className, clearable, onClear, type, ...rest
  } = props;

  const [type2, setType2] = useState('text');

  useEffect(() => {
    if (type) setType2(type);
  }, [type]);

  function clear() {
    const { current } = inputElement;
    if (!current) return;
    current.value = '';
    if (onClear) onClear();
    current.focus();
  }

  function showClearIcon() {
    if (!clearable) return null;
    const { current } = inputElement;
    if (!current) return null;
    if (!current.value) return null;
    return (
      <span title="clear value" className={baseClass('icon')}>
        <Icon name="clear" className={baseClass('icon', 'clear')} onClick={clear} />
      </span>
    );
  }

  function showPassword() {
    setType2('text');
  }

  function hidePassword() {
    setType2('password');
  }

  function showPasswordIcon() {
    if (type !== 'password') return null;
    if (type2 === 'password') {
      return (
        <span title="show password" className={baseClass('icon')}>
          <Icon name="show" className={baseClass('icon', 'password')} onClick={showPassword} />
        </span>
      );
    }
    return (
      <span title="hide password" className={baseClass('icon')}>
        <Icon name="hide" className={baseClass('icon', 'password')} onClick={hidePassword} />
      </span>
    );
  }

  return (
    <div className={baseClass('wrapper')}>
      <input
        ref={inputElement}
        className={classes(baseClass(), className)}
        type={type2}
        {...rest}
      />
      <div className={baseClass('icons')}>
        {showPasswordIcon()}
        {showClearIcon()}
      </div>
    </div>
  );
};


export default Input;

export { Input };
