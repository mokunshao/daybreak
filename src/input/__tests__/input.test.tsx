import { render } from '@testing-library/react';
import React from 'react';
import { Input } from '../input';

describe('Input', () => {
  it('exit', () => {
    expect(Input).toBeTruthy();
    expect(<Input />).toBeTruthy();
  });
  it('match snapshot', () => {
    const { asFragment } = render(<Input />);
    expect(asFragment()).toMatchSnapshot();
  });
});
