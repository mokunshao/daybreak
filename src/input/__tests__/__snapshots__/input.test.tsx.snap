// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input match snapshot 1`] = `
<DocumentFragment>
  <div
    class="daybreak-input-wrapper"
  >
    <input
      class="daybreak-input"
      type="text"
    />
    <div
      class="daybreak-input-icons"
    />
  </div>
</DocumentFragment>
`;
